weed/s3api/auth_signature_v4.go

```diff
@@ -154,248 +157,246 @@ func parseSignV4(v4Auth string) (sv signValues, aec s3err.ErrorCode) {
-	errCode = iam.verifySignatureWithPath(extractedSignedHeaders, hashedPayload, queryStr, req.URL.Path, req.Method, foundCred.SecretKey, t, signV4Values)
-	if errCode == s3err.ErrNone {
-		return identity, errCode
+	// 9. Verify with the original path
+	calculatedSignature, errCode = verify(r.URL.Path)
+	if errCode != s3err.ErrNone {
+		return nil, nil, "", nil, errCode
 	}
 
-	return nil, errCode
+	return identity, cred, calculatedSignature, authInfo, s3err.ErrNone
 }
 
-// buildPathWithForwardedPrefix combines forwarded prefix with URL path while preserving trailing slashes.
-// This ensures compatibility with S3 SDK signatures that include trailing slashes for directory operations.
-func buildPathWithForwardedPrefix(forwardedPrefix, urlPath string) string {
-	fullPath := forwardedPrefix + urlPath
-	hasTrailingSlash := strings.HasSuffix(urlPath, "/") && urlPath != "/"
-	cleanedPath := path.Clean(fullPath)
-	if hasTrailingSlash && !strings.HasSuffix(cleanedPath, "/") {
-		cleanedPath += "/"
-	}
-	return cleanedPath
-}
-
-// verifySignatureWithPath verifies signature with a given path (used for both normal and prefixed paths).
-func (iam *IdentityAccessManagement) verifySignatureWithPath(extractedSignedHeaders http.Header, hashedPayload, queryStr, urlPath, method, secretKey string, t time.Time, signV4Values signValues) s3err.ErrorCode {
-	// Get canonical request.
-	canonicalRequest := getCanonicalRequest(extractedSignedHeaders, hashedPayload, queryStr, urlPath, method)
-
-	// Get string to sign from canonical request.
-	stringToSign := getStringToSign(canonicalRequest, t, signV4Values.Credential.getScope())
-
-	// Get hmac signing key.
-	signingKey := getSigningKey(secretKey, signV4Values.Credential.scope.date.Format(yyyymmdd), signV4Values.Credential.scope.region, signV4Values.Credential.scope.service)
-
-	// Calculate signature.
+// calculateAndVerifySignature contains the core logic for creating the canonical request,
+// string-to-sign, and comparing the final signature.
+func calculateAndVerifySignature(secretKey, method, urlPath, queryStr string, extractedSignedHeaders http.Header, authInfo *v4AuthInfo) (string, s3err.ErrorCode) {
+	canonicalRequest := getCanonicalRequest(extractedSignedHeaders, authInfo.HashedPayload, queryStr, urlPath, method)
+	stringToSign := getStringToSign(canonicalRequest, authInfo.Date, authInfo.Scope)
+	signingKey := getSigningKey(secretKey, authInfo.Date.Format(yyyymmdd), authInfo.Region, authInfo.Service)
 	newSignature := getSignature(signingKey, stringToSign)
 
-	// Verify if signature match.
-	if !compareSignatureV4(newSignature, signV4Values.Signature) {
-		return s3err.ErrSignatureDoesNotMatch
+	if !compareSignatureV4(newSignature, authInfo.Signature) {
+		glog.V(4).Infof("Signature mismatch. Details:\n- Canonical Request: %s\n- String to Sign: %s\n- Calculated Signature: %s\n- Provided Signature: %s",
+			canonicalRequest, stringToSign, newSignature, authInfo.Signature)
```

Comment:
Do note the logging statement here on failure and very verbose output, useful for me debugging without rebuilding locally + deploying a docker image.